tests/func/test_diff.py

```diff
@@ -1,267 +1,274 @@
+    }
+
+    assert dvc.diff("HEAD~1", target="missing") == {
+        "added": [],
+        "deleted": [],
+        "modified": [],
+    }
+
+
+def test_directories(tmp_dir, scm, dvc):
+    tmp_dir.dvc_gen({"dir": {"1": "1", "2": "2"}}, commit="add a directory")
+    tmp_dir.dvc_gen({"dir": {"3": "3"}}, commit="add a file")
+    tmp_dir.dvc_gen({"dir": {"2": "two"}}, commit="modify a file")
+
+    (tmp_dir / "dir" / "2").unlink()
+    dvc.add("dir")
+    scm.add("dir.dvc")
+    scm.commit("delete a file")
+
+    assert dvc.diff(":/init", ":/directory") == {
+        "added": [
+            {
+                "filename": os.path.join("dir", ""),
+                "checksum": "5fb6b29836c388e093ca0715c872fe2a.dir",
+            },
+            {"filename": os.path.join("dir", "1"), "checksum": digest("1")},
+            {"filename": os.path.join("dir", "2"), "checksum": digest("2")},
+        ],
+        "deleted": [],
+        "modified": [],
+    }
+
+    assert dvc.diff(":/directory", ":/modify") == {
+        "added": [
+            {"filename": os.path.join("dir", "3"), "checksum": digest("3")}
+        ],
+        "deleted": [],
+        "modified": [
+            {
+                "filename": os.path.join("dir", ""),
+                "checksum": {
+                    "old": "5fb6b29836c388e093ca0715c872fe2a.dir",
+                    "new": "9b5faf37366b3370fd98e3e60ca439c1.dir",
+                },
+            },
+            {
+                "filename": os.path.join("dir", "2"),
+                "checksum": {"old": digest("2"), "new": digest("two")},
+            },
+        ],
+    }
+
+    assert dvc.diff(":/modify", ":/delete") == {
+        "added": [],
+        "deleted": [
+            {"filename": os.path.join("dir", "2"), "checksum": digest("two")}
+        ],
+        "modified": [
+            {
+                "filename": os.path.join("dir", ""),
+                "checksum": {
+                    "old": "9b5faf37366b3370fd98e3e60ca439c1.dir",
+                    "new": "83ae82fb367ac9926455870773ff09e6.dir",
+                },
+            }
+        ],
+    }
+
+
+def test_json(tmp_dir, scm, dvc, mocker, capsys):
+    assert 0 == main(["diff", "--json"])
+    assert not capsys.readouterr().out
+
+    diff = {
+        "added": [{"filename": "file", "checksum": digest("text")}],
+        "deleted": [],
+        "modified": [],
+    }
+    mocker.patch("dvc.repo.Repo.diff", return_value=diff)
+    assert 0 == main(["diff", "--json"])
+    assert (
+        json.dumps(
+            {
+                "added": [{"filename": "file", "checksum": digest("text")}],
+                "deleted": [],
+                "modified": [],
+            }
         )
-        msg3 = "+{0} with md5 {1}".format(
-            self.test_dct[diff.DIFF_LIST][0][diff.DIFF_TARGET],
-            self.test_dct[diff.DIFF_LIST][0][diff.DIFF_NEW_CHECKSUM],
+        in capsys.readouterr().out
+    )
+
+
+def test_cli(tmp_dir, scm, dvc, mocker, capsys):
```

Comment:
Let's move this to `tests/unit/command/test_diff.py`, same as we do in other similar commands.

Reply:
Btw, we could test `_format` specifically, just move it into an out-of-class method. The CmdDiff could be tested same way we do it in `tests/unit/command/test_get.py`.

Reply:
neat, @efiop :ok_hand: